File: app/src/main/java/com/tiktokaitool/MainActivity.java
package com.tiktokaitool;

import android.animation.ObjectAnimator;
import android.content.*;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.*;
import android.provider.Settings;
import android.view.View;
import android.view.animation.OvershootInterpolator;
import android.widget.*;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;

public class MainActivity extends AppCompatActivity {
    private static final int OVERLAY_REQ = 1001;
    private EditText etUrl;
    private TextView tvStatus, tvApiWarn;
    private PrefsManager prefs;

    @Override
    protected void onCreate(Bundle b) {
        super.onCreate(b);
        setContentView(R.layout.activity_main);
        prefs     = new PrefsManager(this);
        etUrl     = findViewById(R.id.et_url);
        tvStatus  = findViewById(R.id.tv_status);
        tvApiWarn = findViewById(R.id.tv_api_warn);

        animateCards();

        findViewById(R.id.btn_open_tiktok).setOnClickListener(v -> { animBtn(v); openTikTok(); });
        findViewById(R.id.btn_ai_finder).setOnClickListener(v -> { animBtn(v); startAiFinder(); });
        findViewById(R.id.btn_paste).setOnClickListener(v -> pasteUrl());
        findViewById(R.id.btn_download).setOnClickListener(v -> { animBtn(v); startDownload(); });
        findViewById(R.id.btn_stop_overlay).setOnClickListener(v -> stopOverlay());
        findViewById(R.id.btn_history).setOnClickListener(v ->
            startActivity(new Intent(this, HistoryActivity.class)));
        findViewById(R.id.btn_settings).setOnClickListener(v ->
            startActivity(new Intent(this, SettingsActivity.class)));
        findViewById(R.id.tv_api_warn).setOnClickListener(v ->
            startActivity(new Intent(this, SettingsActivity.class)));
    }

    private void animateCards() {
        int[] ids = {R.id.card_header, R.id.card_actions, R.id.card_download, R.id.card_bottom};
        for (int i = 0; i < ids.length; i++) {
            View v = findViewById(ids[i]);
            if (v == null) continue;
            v.setAlpha(0f); v.setTranslationY(60f);
            v.animate().alpha(1f).translationY(0f)
                .setDuration(400).setStartDelay(i * 80L)
                .setInterpolator(new android.view.animation.DecelerateInterpolator())
                .start();
        }
    }

    private void animBtn(View v) {
        v.animate().scaleX(0.93f).scaleY(0.93f).setDuration(80)
            .withEndAction(() -> v.animate().scaleX(1f).scaleY(1f)
                .setDuration(180).setInterpolator(new OvershootInterpolator()).start())
            .start();
    }

    private void openTikTok() {
        String[] pkgs = {"com.zhiliaoapp.musically", "com.ss.android.ugc.trill"};
        for (String pkg : pkgs) {
            try {
                Intent i = getPackageManager().getLaunchIntentForPackage(pkg);
                if (i != null) { startActivity(i); return; }
            } catch (Exception ignored) {}
        }
        try { startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=com.zhiliaoapp.musically"))); }
        catch (Exception e) { startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=com.zhiliaoapp.musically"))); }
    }

    private void startAiFinder() {
        if (prefs.getApiKey().isEmpty()) {
            Toast.makeText(this, "Set Claude API key in Settings first!", Toast.LENGTH_LONG).show();
            startActivity(new Intent(this, SettingsActivity.class)); return;
        }
        if (!Settings.canDrawOverlays(this)) {
            new AlertDialog.Builder(this)
                .setTitle("Overlay Permission Needed")
                .setMessage("TikAI needs to draw over other apps to show the FIND button while you browse TikTok.")
                .setPositiveButton("Grant", (d, w) -> startActivityForResult(
                    new Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION,
                        Uri.parse("package:" + getPackageName())), OVERLAY_REQ))
                .setNegativeButton("Cancel", null).show();
            return;
        }
        launchOverlayAndTikTok();
    }

    private void launchOverlayAndTikTok() {
        Intent si = new Intent(this, OverlayService.class);
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) startForegroundService(si);
        else startService(si);
        new Handler().postDelayed(this::openTikTok, 500);
        Toast.makeText(this, "FIND button active! Drag it anywhere.", Toast.LENGTH_LONG).show();
        updateStatus();
    }

    private void stopOverlay() {
        Intent i = new Intent(this, OverlayService.class);
        i.setAction("STOP"); startService(i);
        new Handler().postDelayed(this::updateStatus, 400);
    }

    private void pasteUrl() {
        ClipboardManager cm = (ClipboardManager) getSystemService(CLIPBOARD_SERVICE);
        if (cm != null && cm.hasPrimaryClip() && cm.getPrimaryClip() != null) {
            CharSequence t = cm.getPrimaryClip().getItemAt(0).getText();
            if (t != null && (t.toString().contains("tiktok.com") || t.toString().contains("vm.tiktok"))) {
                etUrl.setText(t.toString());
                Toast.makeText(this, "Pasted!", Toast.LENGTH_SHORT).show();
            } else Toast.makeText(this, "No TikTok URL in clipboard", Toast.LENGTH_SHORT).show();
        }
    }

    private void startDownload() {
        String url = etUrl.getText().toString().trim();
        if (url.isEmpty()) { Toast.makeText(this, "Paste a TikTok URL first", Toast.LENGTH_SHORT).show(); return; }
        if (!url.contains("tiktok")) { Toast.makeText(this, "Not a valid TikTok URL", Toast.LENGTH_SHORT).show(); return; }
        startActivity(new Intent(this, DownloadsActivity.class).putExtra("url", url));
    }

    private void updateStatus() {
        boolean active = OverlayService.isRunning;
        tvStatus.setText(active ? "● AI Overlay Active" : "● AI Overlay Off");
        tvStatus.setTextColor(getResources().getColor(active ? R.color.green : R.color.text_secondary));
        findViewById(R.id.btn_stop_overlay).setVisibility(active ? View.VISIBLE : View.GONE);
    }

    @Override
    protected void onResume() {
        super.onResume();
        updateStatus();
        tvApiWarn.setVisibility(prefs.getApiKey().isEmpty() ? View.VISIBLE : View.GONE);
    }

    @Override
    protected void onActivityResult(int req, int res, Intent data) {
        super.onActivityResult(req, res, data);
        if (req == OVERLAY_REQ && Settings.canDrawOverlays(this)) launchOverlayAndTikTok();
    }
}           }

File: app/src/main/java/com/tiktokaitool/ui/MainActivity.kt
package com.tiktokaitool.ui

import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.tiktokaitool.R
import com.tiktokaitool.databinding.ActivityMainBinding
import com.tiktokaitool.service.OverlayService
import com.tiktokaitool.utils.PrefsManager
import com.tiktokaitool.utils.TikTokUtils

class MainActivity : AppCompatActivity() {

    private lateinit var binding: ActivityMainBinding
    private lateinit var prefs: PrefsManager

    companion object {
        private const val OVERLAY_PERMISSION_CODE = 1001
        private const val TIKTOK_PACKAGE_MUSICALLY = "com.zhiliaoapp.musically"
        private const val TIKTOK_PACKAGE_TRILL = "com.ss.android.ugc.trill"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityMainBinding.inflate(layoutInflater)
        setContentView(binding.root)
        prefs = PrefsManager(this)
        setupUI()
        checkApiKey()
    }

    private fun setupUI() {
        binding.btnOpenTiktok.setOnClickListener { openTikTok() }

        binding.btnAiFinder.setOnClickListener {
            if (prefs.getApiKey().isBlank()) {
                showApiKeyDialog(); return@setOnClickListener
            }
            if (!Settings.canDrawOverlays(this)) {
                requestOverlayPermission(); return@setOnClickListener
            }
            openTikTokWithOverlay()
        }

        binding.btnDownloads.setOnClickListener {
            startActivity(Intent(this, DownloadsActivity::class.java))
        }
        binding.btnSettings.setOnClickListener {
            startActivity(Intent(this, SettingsActivity::class.java))
        }

        binding.btnPasteDownload.setOnClickListener {
            val cm = getSystemService(CLIPBOARD_SERVICE) as android.content.ClipboardManager
            val clip = cm.primaryClip?.getItemAt(0)?.text?.toString() ?: ""
            if (clip.contains("tiktok.com") || clip.contains("vm.tiktok")) {
                binding.etVideoUrl.setText(clip)
                Toast.makeText(this, "Pasted!", Toast.LENGTH_SHORT).show()
            } else {
                Toast.makeText(this, "No TikTok link in clipboard", Toast.LENGTH_SHORT).show()
            }
        }

        binding.btnDownloadVideo.setOnClickListener {
            val url = binding.etVideoUrl.text.toString().trim()
            if (url.isNotBlank()) startDownload(url)
            else Toast.makeText(this, "Enter a TikTok URL first", Toast.LENGTH_SHORT).show()
        }

        updateOverlayStatus()
    }

    private fun openTikTokWithOverlay() {
        startOverlayService()
        Handler(Looper.getMainLooper()).postDelayed({ openTikTok() }, 400)
        Toast.makeText(this, "FIND button is active on the left side of TikTok!", Toast.LENGTH_LONG).show()
    }

    private fun openTikTok() {
        val pm = packageManager
        val pkg = when {
            isAppInstalled(pm, TIKTOK_PACKAGE_MUSICALLY) -> TIKTOK_PACKAGE_MUSICALLY
            isAppInstalled(pm, TIKTOK_PACKAGE_TRILL) -> TIKTOK_PACKAGE_TRILL
            else -> null
        }
        if (pkg != null) {
            val intent = pm.getLaunchIntentForPackage(pkg)?.apply {
                addFlags(Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_ACTIVITY_RESET_TASK_IF_NEEDED)
            }
            if (intent != null) startActivity(intent)
            else Toast.makeText(this, "Could not launch TikTok", Toast.LENGTH_SHORT).show()
        } else {
            try { startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("market://details?id=$TIKTOK_PACKAGE_MUSICALLY"))) }
            catch (e: Exception) { startActivity(Intent(Intent.ACTION_VIEW, Uri.parse("https://play.google.com/store/apps/details?id=$TIKTOK_PACKAGE_MUSICALLY"))) }
        }
    }

    private fun startOverlayService() {
        if (Settings.canDrawOverlays(this))
            startForegroundService(Intent(this, OverlayService::class.java))
    }

    private fun startDownload(url: String) {
        if (!TikTokUtils.isValidTikTokUrl(url)) {
            Toast.makeText(this, "Invalid TikTok URL", Toast.LENGTH_SHORT).show(); return
        }
        startActivity(Intent(this, DownloadsActivity::class.java).putExtra("download_url", url))
    }

    private fun requestOverlayPermission() {
        MaterialAlertDialogBuilder(this)
            .setTitle("Permission Required")
            .setMessage("TikAI needs Display over other apps permission to show the FIND button while you watch TikTok.")
            .setPositiveButton("Grant") { _, _ ->
                startActivityForResult(
                    Intent(Settings.ACTION_MANAGE_OVERLAY_PERMISSION, Uri.parse("package:$packageName")),
                    OVERLAY_PERMISSION_CODE)
            }
            .setNegativeButton("Cancel", null).show()
    }

    private fun showApiKeyDialog() {
        startActivity(Intent(this, SettingsActivity::class.java).putExtra("focus_api_key", true))
        Toast.makeText(this, "Set your Claude API key first", Toast.LENGTH_LONG).show()
    }

    private fun checkApiKey() {
        binding.cardApiWarning.visibility = if (prefs.getApiKey().isBlank()) View.VISIBLE else View.GONE
    }

    private fun updateOverlayStatus() {
        val active = Settings.canDrawOverlays(this) && OverlayService.isRunning
        binding.tvOverlayStatus.text = if (active) "AI Overlay: Active" else "AI Overlay: Inactive"
        binding.tvOverlayStatus.setTextColor(ContextCompat.getColor(this,
            if (active) R.color.green_active else R.color.text_secondary))
    }

    private fun isAppInstalled(pm: PackageManager, pkg: String) = try {
        pm.getPackageInfo(pkg, 0); true
    } catch (e: PackageManager.NameNotFoundException) { false }

    override fun onResume() { super.onResume(); updateOverlayStatus(); checkApiKey() }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        super.onActivityResult(requestCode, resultCode, data)
        if (requestCode == OVERLAY_PERMISSION_CODE) {
            if (Settings.canDrawOverlays(this)) openTikTokWithOverlay()
            else Toast.makeText(this, "Permission denied", Toast.LENGTH_SHORT).show()
        }
    }
}